components/suggest/src/db.rs

```diff
@@ -197,58 +206,106 @@ impl<'a> SuggestDao<'a> {
+                            score,
+                        })
                     }
-                    SuggestionProvider::Amo => {
-                        let full_suffix = row.get::<_, String>("keyword_suffix")?;
+                )
+            }
+            )?;
+        Ok(suggestions)
+    }
+
+    pub fn fetch_wikipedia_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
+        let keyword_lowercased = &query.keyword.to_lowercase();
+        let suggestions = self.conn.query_rows_and_then_cached(
+            "SELECT s.id, k.rank, s.title, s.url
+             FROM suggestions s
+             JOIN keywords k ON k.suggestion_id = s.id
+             WHERE s.provider = :provider AND
+             k.keyword = :keyword",
+            named_params! {
+                ":keyword": keyword_lowercased,
+                ":provider": SuggestionProvider::Wikipedia
+            },
+            |row| -> Result<Suggestion> {
+                let suggestion_id: i64 = row.get("id")?;
+                let title = row.get("title")?;
+                let raw_url = row.get::<_, String>("url")?;
+
+                let keywords: Vec<String> = self.conn.query_rows_and_then_cached(
+                    "SELECT keyword FROM keywords
+                     WHERE suggestion_id = :suggestion_id AND rank >= :rank
+                     ORDER BY rank ASC",
+                    named_params! {
+                        ":suggestion_id": suggestion_id,
+                        ":rank": row.get::<_, i64>("rank")?,
+                    },
+                    |row| row.get(0),
+                )?;
+                let icon = self.conn.try_query_one(
+                    "SELECT i.data
+                     FROM icons i
+                     JOIN wikipedia_custom_details s ON s.icon_id = i.id
+                     WHERE s.suggestion_id = :suggestion_id",
+                    named_params! {
+                        ":suggestion_id": suggestion_id
+                    },
+                    true,
+                )?;
+                Ok(Suggestion::Wikipedia {
+                    title,
+                    url: raw_url,
+                    full_keyword: full_keyword(keyword_lowercased, &keywords),
+                    icon,
+                })
+            },
+        )?;
+        Ok(suggestions)
+    }
+
+    pub fn fetch_amo_suggestions(&self, query: &SuggestionQuery) -> Result<Vec<Suggestion>> {
+        let keyword_lowercased = &query.keyword.to_lowercase();
+        let (keyword_prefix, keyword_suffix) = split_keyword(keyword_lowercased);
+        let suggestions = self.conn.query_rows_and_then_cached(
+            "SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, k.confidence, k.keyword_suffix
+                     FROM suggestions s
+                     JOIN prefix_keywords k ON k.suggestion_id = s.id
+                     WHERE k.keyword_prefix = :keyword_prefix AND s.provider = :provider",
```

Comment:
I was unsure whether it's worth using `ORDER BY` and `LIMIT` here and the other select statements to to reduce the number of suggestions we have to sort after. I don't think it's worth it for AMP since we're matching on keyword/keyword partials and that should only return 1 suggestion, but for AMO and Pocket that have prefixes...

Reply:
Thanks for highlighting this! Here are our top prefixes for AMO and Pocket, and the number of times they're repeated:

```
sqlite> SELECT keyword_prefix, count(*) FROM prefix_keywords GROUP BY keyword_prefix HAVING count(*) > 1 ORDER BY 2 DESC LIMIT 5;
how|209
best|56
the|37
vegan|37
what|35
```

...So doing `ORDER BY` and `LIMIT` in these queries sounds like a great idea to me! Using the original `query.limit` feels like a good starting point: if the caller passes a limit of `1`, even in the most impossible case of each provider returning 1 suggestion, that's still 5 suggestions at most, that we'll then truncate to 1.

If even that turns out to be inefficient, we can try dynamically adjusting the limit for each provider, based on the number of suggestions remaining, and the priority of the provider—but starting with the original limit still sounds good to me!